Create KEP for deprecating and removing SelfLink

wojtek-t opened the pull request:
/assign @lavalamp @deads2k @liggitt @smarterclayton 

@justinsb - something we used to talk in the past about to

wojtek-t commented on keps/sig-api-machinery/20190711-remove-selflink.md:
```diff
@@ -0,0 +1,169 @@
+---
+title: Deprecate and remove SelfLink
```
The KEP requires extending it visibly, but I would like to first agree that this is something we even want to proceed with.

liggitt commented:
I'm generally in favor of this. It would be good to scan for uses of selflink (in kubernetes/kubernetes and in other https://github.com/kubernetes-client libraries) that would be affected (for example, https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/reference/ref.go#L76-L97)

smarterclayton commented:
We would probably want to offer in the client library a way or description of how to recover this (if they needed).  I.e. req.URL.Path in Go is the same if they aren't behind an intercepting proxy.

wojtek-t commented:
> I'm generally in favor of this. It would be good to scan for uses of selflink (in kubernetes/kubernetes and in other https://github.com/kubernetes-client libraries) that would be affected (for example, https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/reference/ref.go#L76-L97)

@liggitt - thanks.
I added a list of all places that can be related to the KEP - the one that you pointed (in addition to returning it as part of Table format) seems to be the only worth-considering one.
But I think that GV of an object should always return non-empty version, so I'm not sure if that case is really a needed one?
But if you have any thoughts about it, I would appreciate it.

> I'm in favor of this as well, agree with jordan's points.
We would probably want to offer in the client library a way or description of how to recover this (if they needed). I.e. req.URL.Path in Go is the same if they aren't behind an intercepting proxy.

Adding a sentence of documentation is probably trivial (though for me it's not obvious where exactly it should be added). As for "recovering" it, once we remove this field, what do you really mean by that?

lavalamp commented:
I think selfLink leaves out exactly the thing that you can't tell from looking at a stored object, which is what *cluster/server* it came from. Everything else can be computed. Although I don't think it's good for us to force clients to do that computation, I don't think precomputing for them like this is a useful thing to do, either, because it doesn't help when you have a new object.

I looked in old sources the other day and this appears to be a holdover from the very very first draft of the API types, from before I regularized them into "JSONBase" (which turned into today's ObjectMeta); I'm not sure it got anywhere near as much design attention as other parts of metadata.

Anyway, I am fine with removing it (with a long deprecation period :/ ), but I also don't see the pressing need for doing it right now?

I think a "location" or "source cluster" sort of field would be more useful, and it might be good to introduce at the same time since the same code paths would be needed to set something like this. Of course this would need its own justification, probably.

wojtek-t commented:
@lavalamp 
- I added the part that it leaves out the cluster it comes from to the KEP (it's useful to have it stated explicitly, thanks)
- I also added "adding "source-cluster"" explicitly as non-goals for this KEP - I don't think we should merge those two things (I'm actually not even convinced that the code-paths would be similar here, I think this could be done ina simpler way)

> Anyway, I am fine with removing it (with a long deprecation period :/ ),

great!

> but I also don't see the pressing need for doing it right now?

With the deprecation period that we need to for it, if there would be a pressing need for it, it would be too late. I would much rather start this process of deprecation now to be in a position where we can remove it in ~an year from now. Also - I think that from the SIG would require basically approving this KEP, so I think it's not much of work...

lavalamp commented:
I'm fine with deprecating this ~now, with removal in a year unless we find some substantial reason to keep it.

wojtek-t commented:
@lavalamp @smarterclayton @liggitt - I cleaned up and updated the KEP to fill (hopefully) all necessary information.
PTAL to see if that plan seems reasonable and we could proceed with it.

lavalamp commented:
@deads2k wanted to make a request here (at least one release where it can be turned back on via flag), we're waiting for that and then we'll approve.

wojtek-t commented:
@lavalamp @deads2k - I updated the KEP that:
- we will also introduce a feature gate and hide the logic setting SelfLink behind it
- in 1.20, we only change the value of feature gate to no set it
- only in 1.21 we will get rid of the field entirely

Let me know if that seems reasonable.

deads2k commented on keps/sig-api-machinery/20190711-remove-selflink.md:
```diff
@@ -0,0 +1,202 @@
+We will also introduce a feature gate to allow disabling setting `SelfLink` fields
+and opaque the logic setting it behind this feature gate.
+
+In v1.20 (12 months and 4 release from v1.16) we will switch off the feature gate
```
Thanks, this is what I was looking for.

deads2k commented:
/lgtm
/hold

holding so @lavalamp can have a second look.

lavalamp commented:
/approve
/lgtm
/hold cancel
